File: Python/maximum subarray.py
a =int(input())
num_list = []

for i in range(a):
    s = int(input())
    num_list.append(s)


def find_max_sub(arr,low,high):
    if high == low:
        return (low,high,arr[low])
    else:
        middle = int((low+high)/2)
        (left_low,left_high,left_sum) = find_max_sub(arr,low,middle)
        (right_low, right_high, right_sum) = find_max_sub(arr, middle+1,high)
        (cross_low,cross_high,cross_sum)= find_max_cross_sub(arr, low, middle, high)

        if left_sum >= right_sum and left_sum >= cross_sum:
            return (left_low,left_high,left_sum)
        elif right_sum >= left_sum and right_sum >= cross_sum:
            return (right_low, right_high, right_sum)
        else: return (cross_low,cross_high,cross_sum)
def find_max_cross_sub(arr, low, middle, high):
    left_sum = -999999
    sum =0
    for i in range(middle,low-1,-1):
        sum = sum + arr[i]
        if sum > left_sum:
            left_sum = sum
            max_left = i
    right_sum = -999999
    sum =0
    for j in range(middle+1, high+1):
        sum = sum + arr[j]
        if sum > right_sum:
            right_sum = sum
            max_right = j
    return (max_left,max_right,left_sum+right_sum)

x = find_max_sub(num_list,0,a-1)
for i in range(3):
    print(x[i])

File: C/maximum subarray.cpp
#include <stdio.h>
#include <stdlib.h> 
int find_max_sub_middle(int arr[], int low, int middle, int high) {
	int left_sum = -99999;
	int sum = 0;
	int max_left =0;
	
	for (int i = middle; i >= low; i--) {
		sum = sum + arr[i];
		if (sum > left_sum)
			left_sum = sum;
			max_left = i;
			
	}

	int right_sum = -99999;
	sum = 0;
	int max_right=0;

	for (int j = middle + 1; j <= high; j++) {
		sum = sum + arr[j];
		if (sum > right_sum)
			right_sum = sum;
			max_right = j;
	}

	return (left_sum + right_sum);
}
int find_max_sub(int arr[], int low, int high) {
	if (high == low)
		return arr[low];
	else {
		int middle = (high + low) / 2;

		int max_sum_left = find_max_sub(arr, low, middle);
		int max_sum_right = find_max_sub(arr, middle + 1, high);
		int max_sum_cross = find_max_sub_middle(arr, low, middle, high);

		if (max_sum_left >= max_sum_right && max_sum_left >= max_sum_cross)
			return max_sum_left;
		else if (max_sum_right >= max_sum_left && max_sum_right >= max_sum_cross)
			return max_sum_right;
		else return max_sum_cross;
	}
}


int main() {
	int size;
	scanf("%d", &size);
	int* arr = (int*)malloc(sizeof(int) * size);

	for (int i = 0; i < size; i++) {
		scanf("%d", &arr[i]);
	}
	int ans = find_max_sub(arr,0,size-1);
	
	printf("%d",ans);

	free(arr);

	return 0;
}
